dallinger/experiment_server/experiment_server.py

```diff
@@ -1556,18 +1556,24 @@ def worker_complete():
 
 
 def _worker_complete(participant_id):
-    participants = models.Participant.query.filter_by(id=participant_id).all()
-    if not participants:
+    participant = models.Participant.query.get(participant_id)
+    if not participant:
         raise KeyError()
 
-    participant = participants[0]
+    if participant.end_time is not None:  # Provide idempotence
```

Comment:
I think this is safe because this `_worker_complete` function seems to be called before the `worker_function` which calls various `WorkerEvent` instances (like `AssignmentSubmitted`) which also set `end_time` via the `update_particant_end_time` method (note typo). However, I'm a little wary that we're still setting/updating this in multiple places.

Reply:
Yeah, this is a mess because (perhaps among other reasons) MTurk participants are a special case for which we defer all the `WorkerEvent` tasks until word comes in from MTurk that the HIT has been submitted. I'm more worried about this the more I think about it (some experiments calculate bonuses based in part on `end_time`) but I'm not sure it makes a difference for this PR.

Reply:
Yeah, this seems like it should always be OK since it's always fired before any of those `WorkerEvent` things get called. The whole things seems a little brittle to me, especially if we're using end_time for anything important like bonus payments.

Reply:
I'm not sure why the single `end_time` setting at the beginning of `_worker_complete()` couldn't be the one and only place this happens, but we'll probably need to do give it a thorough review.